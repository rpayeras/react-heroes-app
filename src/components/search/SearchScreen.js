import React, { useMemo } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { getHeroByName } from "../../selectors/hero";
import { HeroCard } from "../hero/HeroCard";
import queryString from "query-string";

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q: queryUrl = "" } = queryString.parse(location.search);

  const [{ q }, handleInputChange] = useForm({
    q: queryUrl,
  });

  const heroes = useMemo(() => getHeroByName(queryUrl), [queryUrl]);

  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${q}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-4">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Search a hero..."
              className="form-control"
              name="q"
              autoComplete="off"
              onChange={handleInputChange}
              value={q}
            />
            <button
              className="btn btn-outline-primary mt-1 btn-block"
              type="submit"
            >
              Search
            </button>
          </form>
        </div>

        <div className="col-8">
          {q === "" ? (
            <div className="alert alert-info">Search a hero</div>
          ) : (
            heroes.length === 0 && (
              <div className="alert alert-danger">Sin resultados: {q}</div>
            )
          )}
          {heroes.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
