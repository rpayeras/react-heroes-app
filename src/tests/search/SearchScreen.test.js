import { mount } from "enzyme"
import { MemoryRouter, useNavigate } from "react-router-dom"
import { SearchScreen } from "../../components/search/SearchScreen"

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

describe('Testing SearchScreen component', () => {

    test('should show properly', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        )
         expect(wrapper).toMatchSnapshot()
         expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero')
    })

    test('should show batman chacaracter with queryString value', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect(wrapper.find('input').prop('value')).toBe('batman')
    })

    test('should and error if hero not exists', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmansdsd']}>
                <SearchScreen />
            </MemoryRouter>
        )

        expect(wrapper.find('.alert-danger').exists()).toBe(true)
    })


    test('should write a query and return the result page', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', {
            target: {
                name: 'q',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        })

                
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman')
    })
})