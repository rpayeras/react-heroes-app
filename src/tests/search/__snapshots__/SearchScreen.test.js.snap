// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchScreen component should show properly 1`] = `
Array [
  <h1>
    Search
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="q"
          onChange={[Function]}
          placeholder="Search a hero..."
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1 btn-block"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-8"
    >
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Superman"
                className="card-img-top"
                src="/assets/dc-superman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Superman
                </h5>
                <p
                  className="card-text"
                >
                  Kal-El
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-superman"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Flash"
                className="card-img-top"
                src="/assets/dc-flash.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Flash
                </h5>
                <p
                  className="card-text"
                >
                  Jay Garrick
                </p>
                <p
                  className="text-muted"
                >
                  Jay Garrick, Barry Allen, Wally West, Bart Allen
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-flash"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Green Lantern"
                className="card-img-top"
                src="/assets/dc-green.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Green Lantern
                </h5>
                <p
                  className="card-text"
                >
                  Alan Scott
                </p>
                <p
                  className="text-muted"
                >
                  Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-green"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Green Arrow"
                className="card-img-top"
                src="/assets/dc-arrow.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Green Arrow
                </h5>
                <p
                  className="card-text"
                >
                  Oliver Queen
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-arrow"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Wonder Woman"
                className="card-img-top"
                src="/assets/dc-wonder.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Wonder Woman
                </h5>
                <p
                  className="card-text"
                >
                  Princess Diana
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-wonder"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Martian Manhunter"
                className="card-img-top"
                src="/assets/dc-martian.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Martian Manhunter
                </h5>
                <p
                  className="card-text"
                >
                  J'onn J'onzz
                </p>
                <p
                  className="text-muted"
                >
                  Martian Manhunter
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-martian"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Robin/Nightwing"
                className="card-img-top"
                src="/assets/dc-robin.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Robin/Nightwing
                </h5>
                <p
                  className="card-text"
                >
                  Dick Grayson
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-robin"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Blue Beetle"
                className="card-img-top"
                src="/assets/dc-blue.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Blue Beetle
                </h5>
                <p
                  className="card-text"
                >
                  Dan Garret
                </p>
                <p
                  className="text-muted"
                >
                  Dan Garret, Ted Kord, Jaime Reyes
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-blue"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Black Canary"
                className="card-img-top"
                src="/assets/dc-black.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Black Canary
                </h5>
                <p
                  className="card-text"
                >
                  Dinah Drake
                </p>
                <p
                  className="text-muted"
                >
                  Dinah Drake, Dinah Lance
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/dc-black"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Spider Man"
                className="card-img-top"
                src="/assets/marvel-spider.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Spider Man
                </h5>
                <p
                  className="card-text"
                >
                  Peter Parker
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-spider"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Captain America"
                className="card-img-top"
                src="/assets/marvel-captain.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Captain America
                </h5>
                <p
                  className="card-text"
                >
                  Steve Rogers
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-captain"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Iron Man"
                className="card-img-top"
                src="/assets/marvel-iron.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Iron Man
                </h5>
                <p
                  className="card-text"
                >
                  Tony Stark
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-iron"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Thor"
                className="card-img-top"
                src="/assets/marvel-thor.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Thor
                </h5>
                <p
                  className="card-text"
                >
                  Thor Odinson
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-thor"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Hulk"
                className="card-img-top"
                src="/assets/marvel-hulk.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Hulk
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Banner
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-hulk"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Wolverine"
                className="card-img-top"
                src="/assets/marvel-wolverine.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Wolverine
                </h5>
                <p
                  className="card-text"
                >
                  James Howlett
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-wolverine"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Daredevil"
                className="card-img-top"
                src="/assets/marvel-daredevil.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Daredevil
                </h5>
                <p
                  className="card-text"
                >
                  Matthew Michael Murdock
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-daredevil"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Hawkeye"
                className="card-img-top"
                src="/assets/marvel-hawkeye.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Hawkeye
                </h5>
                <p
                  className="card-text"
                >
                  Clinton Francis Barton
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-hawkeye"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Cyclops"
                className="card-img-top"
                src="/assets/marvel-cyclops.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Cyclops
                </h5>
                <p
                  className="card-text"
                >
                  Scott Summers
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-cyclops"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Silver Surfer"
                className="card-img-top"
                src="/assets/marvel-silver.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Silver Surfer
                </h5>
                <p
                  className="card-text"
                >
                  Norrin Radd
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  />
                </p>
                <a
                  href="/hero/marvel-silver"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
